### PR TITLE
add `deltaSumTimestamp` + docs, tests & minor improvement to `deltaSum`

### PR DESCRIPTION
**Pull request by rf**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

This makes a minor improvement to the `deltaSum` function (that is backwards compatible) and adds a new function called `deltaSumTimestamp`.

The small improvement is to eliminate one of the state bools. I realized that having both `seen_first` and `seen_last` is unnecessary. Since both bools would always have the same value, removing one of them is backwards compatible with old `deltaSum` state data.

The new function, `deltaSumTimestamp`, works similarly but also keeps track of the timestamp of the first value seen and the last value seen. I found that in materialized views that were ordered by a `toStartOfFiveMinute` bucket, the merges were happening in a random order, so the delta sums were all incorrect. By storing the timestamp of the first and last value that is added to a particular `deltaSumTimestamp` state, we can properly order them during merges. This makes it work well in materialized views.

Big thanks to Clickhouse devs for taking my PRs. As always any and all feedback is appreciated.

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- New aggregate function `deltaSumTimestamp` for summing the difference between consecutive rows while maintaining ordering during merge by storing timestamps

### CODE REVIEW

**Comment by rf** [comment]
The fuzzer failures seem to just be flakiness.. if not, if someone with more familiarity with those could point at a particular spot in the logs I should be looking at, I'd appreciate it.

**Comment by Avogar** [comment]
@rt Can you please share scenarios of using deltaSum and deltaSumTimestamp? Now I see that both functions depend on the order and their result will be incorrect when using parallel aggregation (that is enabled as default, but In some cases, you can rely on the order of execution, this applies to cases when SELECT comes from a subquery that uses ORDER BY). There are examples of incorrect execution:
```
:) create table delta (x Int64, time Int64) engine=MergeTree() order by time

CREATE TABLE delta
(
    `x` Int64,
    `time` Int64
)
ENGINE = MergeTree
ORDER BY time

Query id: 361b2a66-3741-40c4-8523-a9f01672cec3

Ok.

0 rows in set. Elapsed: 0.043 sec. 

:) insert into delta select number, number from numbers(100000000)

INSERT INTO delta SELECT
    number,
    number
FROM numbers(100000000)

Query id: d4444f19-639a-4eb3-902d-9908ae7f5425

Ok.

0 rows in set. Elapsed: 3.876 sec. Processed 100.66 million rows, 805.28 MB (25.97 million rows/s., 207.75 MB/s.) 

:) set max_threads=2

SET max_threads = 2

Query id: 1b38c9a1-29d8-44ed-83e8-2d979f287c2c

Ok.

0 rows in set. Elapsed: 0.003 sec. 

:) select deltaSum(x) from delta

SELECT deltaSum(x)
FROM delta

Query id: dc021aff-f600-4672-84cc-639f459ab098

┌─deltaSum(x)─┐
│  2415497401 │
└─────────────┘

1 rows in set. Elapsed: 0.582 sec. Processed 100.00 million rows, 800.00 MB (171.91 million rows/s., 1.38 GB/s.) 

avogar-dev.sas.yp-c.yandex.net :) select deltaSumTimestamp(x, time) from delta

SELECT deltaSumTimestamp(x, time)
FROM delta

Query id: 13512581-2faf-4111-a264-34090a36a308

┌─deltaSumTimestamp(x, time)─┐
│                  686968533 │
└────────────────────────────┘

1 rows in set. Elapsed: 0.990 sec. Processed 100.00 million rows, 1.60 GB (101.03 million rows/s., 1.62 GB/s.) 

```
We are currently in discussion how to deal with functions that depend on the order.
But if such functions are used with data that is ordered by some key (if I understand correctly, deltaSumTimestamp is about this case. When using MergeTree we always have such key), I have an idea how to rewrite such functions so they will work correctly with parallel aggregation (but I'm not sure about it). Let me explain my thoughts on deltaSumTimestamp. Imagine that we have data `x` that is ordered by time `time` and we want to execute `deltaSumTimestamp(x, time)`. Instead of one instance of AggregationFunctionDeltaSumData (lets name it just Data) in our aggregation state we will store a vector of Data. `add` method we almost don't need to change, we just need to know in which of Data in vector we need to insert new value. `merge` method is more interesting, we will merge two vectors like in merge_sort (so we don't merge Data here, only vectors) so that the result vector will be sorted by first_ts. When we call method `insertResultInto` we have already processed all input data and all Data in vector are sorted by fitst_ts, and now we just need to merge all Data in vector into single Data one by one from left to right, so such merge will give us the correct result.
Lets see an example. Imagine we have such data: [(0, 0), (1, 1), (2, 2), (1, 3), (3, 4), (4, 5)] (first is `x`, second is `time`).
And there are three threads, first thread works with [(0, 0), (1, 1)], second with [(2, 2), (1, 3)] and third with [(3, 4), (4, 5)]. After all finished `add` we will have such aggregation states:
[Data(sum = 1, first = 0, last = 1, first_ts = 0, last_ts = 1)] for the first thread (lets name it state1)
[Data(sum = 0, first = 2, last = 1, first_ts = 2, last_ts = 3)] for the second thread (sate2)
[Data(sum = 1, first = 3, last = 4, first_ts = 4, last_ts = 5)] for the third thread (state3)
And imagine we merge it in such sequence: state1, state3, state2:
merge(state1, state3) = [Data(sum = 1, first = 0, last = 1, first_ts = 0, last_ts = 1), Data(sum = 1, first = 3, last = 4, first_ts = 4, last_ts = 5)] = state13
merge(state13, state2) = [Data(sum = 1, first = 0, last = 1, first_ts = 0, last_ts = 1), Data(sum = 0, first = 2, last = 1, first_ts = 2, last_ts = 3), Data(sum = 1, first = 3, last = 4, first_ts = 4, last_ts = 5)] = result_state
And after merging result_state we will get Data(sum = 5, first = 0, last = 4, first_ts = 0, last_ts = 5) and this is correct.
Lets see what happens in current implementation:
merge(state1, state3) = Data(sum = 1 + 1 + (3 - 1) = 4, first = 0, last = 4, first_ts = 0, last_ts = 5) = state13
merge(state13, state2) = (here we have two intervals [0, 5] and [2, 3] which we cannot compare and just ignore rhs state) = Data(sum = 4, first = 0, last = 4, first_ts = 0, last_ts = 5) and this is not correct.
@alexey-milovidov maybe it will be interesting for you and you can share your thoughts.

**Comment by alexey-milovidov** [comment]
We don't have guarantee that one block has one contiguous range of data. So, it may not work.

**Comment by Avogar** [comment]
I will check if it works

**Comment by rf** [comment]
Hey, thanks so much for looking into this!

We are a manufacturing analytics company and we have timeseries data that tracks a continuously increasing number. It will increase up to some maximum, then be reset to zero. You can think of it like a spool of wire with a rotary encoder, that is going to count up to some value, then wind back down to zero. We would like to count the amount the value increased within a specific timeframe; this is the "output" of the process. Here is an example

```sql
create table metrics_test (
    timestamp DateTime CODEC(DoubleDelta),
    name String,
    value Float64 CODEC(Gorilla)
)
ENGINE = MergeTree()
PARTITION BY toYYYYMM(timestamp)
ORDER BY (name, timestamp);

insert into metrics_test (timestamp, name, value) values
('2020-01-04 10:00:00', 'counter', 3),
('2020-01-04 10:00:02', 'counter', 10),
('2020-01-04 10:00:08', 'counter', 30),
('2020-01-04 10:00:10', 'counter', 44),
('2020-01-04 10:00:40', 'counter', 30),
('2020-01-04 10:01:00', 'counter', 0),
('2020-01-04 10:01:01', 'counter', 4),
('2020-01-04 10:01:02', 'counter', 9);

create materialized view metrics_test_1m
ENGINE=AggregatingMergeTree()
PARTITION BY toYYYYMM(bucket) ORDER BY (name, bucket)
POPULATE
AS SELECT toStartOfMinute(timestamp) as bucket, deltaSum(value), name
FROM metrics_test
GROUP BY name, bucket;

select * from metrics_test_1m;

select toStartOfMinute(timestamp) as bucket, name, deltaSum(value),
arraySum(arrayFilter(x -> x > 0, groupArray(runningDifference(value)))) as runningDiff
from metrics_test
group by bucket, name;
```

This works fine with such a small dataset but obviously has problems with our actual data at ingest rate. You can also see that the same result can be achieved by using `runningDifference` function. But that function is not an `AggregateFunction` so we cannot use it in materialized views. We need to have 5m, 30m, 1hr materialized views and would like to have `deltaSum`, hence why I added it as an `AggregateFunction`.

It seems to me that we would need a guarantee that a block contains a contiguous range of data for any implementation of delta sum regardless of whether we store a vector. There may be some way to store less than a full vector, maybe some sort of sample, in order to make this work with various non contiguous pieces of data being aggregated at different times. That may make it an approximation, however.

But primarily I am looking for a solution that will work in a materialized view, where we align the data to some bucket like five or ten minutes. `runningDifference` works fine outside of materialized views. So does `deltaSum` if you select from a subquery with order by.

**Comment by Avogar** [comment]
@rf unfortunately, deltaSumTimestamp and deltaSum won't work properly in materialized view because parallel aggregation is used there. I think I can make deltaSumTimestamp work using my algorithm and using some other tricks that rely on some unreliable invariants, but it would be a dirty solution and we don't want to use it. The only solution I can see now is to not use parallel aggregation when using deltaSum. I will think how to do this.

**Comment by Avogar** [comment]
> Is there no way to check if two rows are actually 'next' to each other according to ordering key?

In general, no way.

> If there was a way to not use parallel aggregation for one field, that would work. 

We can disable parallel aggregation for certain aggregate function. I can do it for deltaSum and you will test if it suits you.

**Comment by Avogar** [comment]
@rf I just noticed, are you sure that you use materialized view with ENGINE=AggregatingMergeTree() correctly? You wrote such query:
```
create materialized view metrics_test_1m
ENGINE=AggregatingMergeTree()
PARTITION BY toYYYYMM(bucket) ORDER BY (name, bucket)
POPULATE
AS SELECT toStartOfMinute(timestamp) as bucket, deltaSum(value), name
FROM metrics_test
GROUP BY name, bucket;
```
But it's wrong, you should use deltaSumState(value) instead of deltaSum(value). When you use AggregatingMergeTree your table should have AggregationFunctionState field to merge it, because otherwise it would work just like MergeTree and two rows with same primary key will be overwritten by one of them. So, you should use such query:
```

**Comment by rf** [comment]
Ah, I was a little hasty in writing that example. Here is an example that is based on what I actually tested:

```
create table default.metrics_stat
(
    bucket      DateTime Codec (DoubleDelta, LZ4),
    machine_id  UUID,
    name        String,

    s_avg       AggregateFunction(avg, Float64),
    s_count     AggregateFunction(count, Float64),
    sum         SimpleAggregateFunction(sum, Float64) Codec (Gorilla),
    min         SimpleAggregateFunction(min, Float64) Codec (Gorilla),
    max         SimpleAggregateFunction(max, Float64) Codec (Gorilla),
    first       SimpleAggregateFunction( any, Float64) Codec (Gorilla),
    last        SimpleAggregateFunction( any, Float64) Codec (Gorilla),
    s_deltasum  AggregateFunction(deltaSum, Float64),
    s_linear_regression AggregateFunction(simpleLinearRegression, UInt32, Float64),
    s_varpop AggregateFunction(varPop, Float64),
) ENGINE = AggregatingMergeTree()
PARTITION BY toYYYYMM(bucket)
ORDER BY (machine_id, name, bucket)
SETTINGS index_granularity = 1024;
```


```
create materialized view default.metrics_stat_view
TO default.metrics_stat
AS
SELECT toStartOfMinute(timestamp)                                                as bucket,
    machine_id,
    name,

    avgState(value)                                                              as s_avg,
    countState(value)                                                            as s_count,
    sum(value)                                                                   as sum,
    min(value)                                                                   as min,
    max(value)                                                                   as max,
    any(value)                                                                   as first,
    anyLast(value)                                                               as last,
    deltaSumState(value)                                                         as s_deltasum,
    simpleLinearRegressionState(toUnixTimestamp(timestamp), value)               as s_linear_regression,
    varPopState(value)                                                           as s_varpop,
FROM metrics
GROUP BY bucket, machine_id, name;
```

And I query from it like this:
```
select bucket, machine_id, name, deltaSumMerge(s_deltasum), sum(sum) from metrics_stat where machine_id='8aedf603-925a-4902-86a8-c12a0450884d' and name='alki' group by bucket, machine_id, name order by bucket;
```

This will produce all zeroes.

**Comment by rf** [comment]
Oh sure, it works fine when used outside of MVs. But when used in a MV:

```
create table default.metrics
(
    timestamp DateTime Codec (DoubleDelta, LZ4),
    name      String,
    value     Float64
) ENGINE = MergeTree()
PARTITION BY toYYYYMM(timestamp)
ORDER BY (name, timestamp);

create table default.metrics_stat
(
    bucket      DateTime Codec (DoubleDelta, LZ4),
    name        String,

    s_deltasum          AggregateFunction(deltaSum, Float64),
    s_deltasumtimestamp AggregateFunction(deltaSumTimestamp, Float64, DateTime)
) ENGINE = MergeTree()
PARTITION BY toYYYYMM(bucket)
ORDER BY (name, bucket);

create materialized view default.metric_stat_view
TO default.metrics_stat
AS
    select toStartOfMinute(timestamp) as bucket,
    name,
    deltaSumStateState(value) as s_deltasumstate,
    deltaSumTimestampState(value, timestamp) as s_deltasumtimestamp
FROM default.metrics
GROUP BY bucket, name;

insert into metrics values (toDateTime('2021-03-17 12:00:00'), 'test', 1);
insert into metrics values (toDateTime('2021-03-17 12:00:01'), 'test', 2);
insert into metrics values (toDateTime('2021-03-17 12:00:02'), 'test', 3);
insert into metrics values (toDateTime('2021-03-17 12:00:03'), 'test', 4);
insert into metrics values (toDateTime('2021-03-17 12:00:04'), 'test', 5);

select bucket, name, deltaSumMerge(s_deltasum), deltaSumTimestampMerge(s_deltasumtimestamp) 
from metrics_stat group by bucket, name order by bucket;

select 
    toStartOfMinute(timestamp) as bucket, name, 
    arraySum(arrayFilter(x -> x > 0, groupArray(runningDifference(value)))) 
from metrics group by bucket, name order by bucket;
```

`deltaSum` will produce 0. I believed that this was because in the MV the points couldn't be ordered which is why I added `deltaSumTimestamp`, which does work.

**Comment by Avogar** [comment]
@rf Now I see the problem deltaSumTimestamp should solve. I didn't get it before, sorry. If `deltaSumTimestamp` can really help you, I will review the code and merge it. But I have and idea how to solve your problem without `deltaSumTimestamp`. Try such query:
```sql
select bucket, name, deltaSumMerge(s_deltasum), deltaSumTimestampMerge(s_deltasumtimestamp) 
from metrics_stat FINAL group by bucket, name order by bucket;
```

Or:

```sql
optimize table metrics_stat final; 
select bucket, name, deltaSumMerge(s_deltasum), deltaSumTimestampMerge(s_deltasumtimestamp) 
from metrics_stat group by bucket, name order by bucket;
```
It seems that such queries can guarantee the right order of merging. Can you try them? If it will work, will you need `deltaSumTimestamp`?

File: src/AggregateFunctions/AggregateFunctionDeltaSum.h
Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionDeltaSum.h:86]: It seems to be incorrect process of counter reset case. Why do we compare `rhs_data->last < place_data->first` instead of `place_data->last > rhs_data->first`? Why do we set `place_data->first = rhs_data->first;` instead of `place_data->last = rhs_data->;`
Lets see some examples:
If we have such sequences: [4, 6] [1, 2] [5, 8]. States are (first=4, last=6, sum=2), (first=1, last=2, sum=1), (first=5, last=8, sum=3).
First merge: merge((first=4, last=6, sum=2), (first=1, last=2, sum=1)) = (first=1, last=6, sum=3), but this state should represent [4, 6, 1, 2] sequence.
Second merge: merge((first=1, last=6, sum=2), (first=5, last=8, sum=3)) = |this case you don't process as count reset, because check 8 < 1 failed and we just skip left result here| = (first=5, last=8, sum=3)

Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionDeltaSum.h:86]: BTW, it seems to be incorrect process of counter reset case. Why do we compare rhs_data->last < place_data->first instead of place_data->last > rhs_data->first? Why do we set place_data->first = rhs_data->first instead of place_data->last = rhs_data->?
Lets see an example:
If we have such sequences: [4, 6] [1, 2] [5, 8]. States are (first=4, last=6, sum=2), (first=1, last=2, sum=1), (first=5, last=8, sum=3).
First merge: merge((first=4, last=6, sum=2), (first=1, last=2, sum=1)) = (first=1, last=6, sum=3), but this state should represent [4, 6, 1, 2] sequence.
Second merge: merge((first=1, last=6, sum=2), (first=5, last=8, sum=3)) = |this case you don't process as count reset, because check 8 < 1 failed and we just skip left result here| = (first=5, last=8, sum=3).

Or I missed something?

**Comment by rf** [comment]
Selecting with `final` or doing `optimize table` doesn't seem to change the results for me. `deltaSumTimestamp` still produces correct results while `deltaSum` does not.

File: src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h
Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h:87]: If we assume that we can have one timestamp multiple times, this code will say that (first_ts=4, last_ts=5) is not before (first_ts=5, last_ts=5). Maybe checking `lhs->last_ts <= rhs->first_ts` will be enough in `before`?

File: src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h
Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h:111]: Empty branch with comment `Do nothing` looks weird. Just add such code before if:
```c++
if (place_data->seen && !rhs_data->seen)
    return;
```

File: src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h
Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h:116]: Why don't we check that `rhs_data->first > place_data->last`?

File: src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h
Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h:124]: Same, we don't check that`place_data->first - rhs_data->last`

**Comment by Avogar** [comment]
@rf I wrote some comments, please, take a look.

File: src/AggregateFunctions/AggregateFunctionDeltaSum.h
Comment (rf) [reply src/AggregateFunctions/AggregateFunctionDeltaSum.h:86]: (Ugh, sorry. I'd had a response to this comment but I forgot to hit "add single comment" so it was in the review which I never submitted.)

Yeah, I think you're right, that is a better comparison to be done here.

Should I fix this in this PR or a separate one?

Comment (Avogar) [reply src/AggregateFunctions/AggregateFunctionDeltaSum.h:86]: > Should I fix this in this PR or a separate one?

I think it would be better to fix in a separate PR with bugfix category to have an ability to backport it

File: src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h
Comment (rf) [reply src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h:87]: This is a good point. However I think that my implementation of `before` is correct, but produces nondeterministic results for equal timestamps. Your suggestion would also produce nondeterministic results. 

I think what we actually need to do is handle the case where the timestamps are equal by comparing the values. We could do this in a final `else` clause of the large `if` statement in `merge`. We would want to either always take the smaller value or the larger value into a new 'singleton' state, where a singleton state is a state where first_ts = last_ts and first = last.

Whether we take the smaller or larger, there is a case where it will lose information, but the sum returned would be deterministic. In particular when there are counter resets around where we have two values with the same timestamp.

If we were to implement before with a simple <=, consider the following singleton states:

```
ts | val
0  | 0
1  | 2
1  | 5
2  | 7
```

If we happen to be merging the (ts=1, val=5) state into the (ts=1, val=2) state, `before` would return true (1 <= 1 is true), so we would say that the state with value 2 happened before the state with value 5. This would produce a new state where (first_ts=1, last_ts=1, first=2, last=5, sum=3). So then we're merging (ts=0, val=0) (first_ts=1, last_ts=1, first=2, last=5, sum=3) (ts=2, val=7). This produces a delta sum of 7.

However, if we happen to merge (ts=1, val=2) into (ts=1, val=5), `before` is still true so we end up with (first_ts=1, last_ts=1, first=5, last=2, sum=0). Then we're merging (ts=0, val=0) (first_ts=1, last_ts=1, first=5, last=2, sum=0) (ts=2, val=7). This produces a delta sum of 10.

So the <= approach is nondeterministic.

I will put up my solution and you can see if it makes sense.

File: src/AggregateFunctions/AggregateFunctionDeltaSum.h
Comment (rf) [reply src/AggregateFunctions/AggregateFunctionDeltaSum.h:86]: Should I also move the other small improvement to a different PR as well?

Comment (Avogar) [reply src/AggregateFunctions/AggregateFunctionDeltaSum.h:86]: We need the new PR only for backporting fixing possible incorrect behaviour, the other small improvement we can leave here

Comment (rf) [reply src/AggregateFunctions/AggregateFunctionDeltaSum.h:86]: https://github.com/ClickHouse/ClickHouse/pull/23437

**Comment by Avogar** [comment]
LGTM, I think we can merge it. But I am thinking about introducing special setting, under which these functions will work, because they are specific and can produce unexpected result in some cases (when parallel aggregation is used).

**Review by Avogar — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@Avogar There is still a lot of tech debt in this code.

1. It is using ALWAYS_INLINE for an unknown reason.
2. Some code uses pointers instead of references.
3. A lot of duplication.
4. More importantly, it compiles for a long time and bloats our binary due to excessive usage of C++ templates.

**Comment by alexey-milovidov** [comment]
I'm addressing it here: https://github.com/ClickHouse/ClickHouse/pull/56350

**Pull request merged**
